packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,152 @@
+  [ButtonSize.XLarge]: IconSize.XLarge,
+};
+
+export type IconType = React.ReactElement<IconProps>;
+
+export { ButtonColor };
+
+export enum ButtonKind {
+  Button = 'button',
+  Link = 'link',
+}
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+  Float = 'tertiaryFloat',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  kind?: ButtonKind;
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (size: ButtonSize, iconOnly: boolean) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(icon.props.className, !iconOnly && 'icon'),
+    });
+};
+
+export type AllowedTags = keyof Pick<JSX.IntrinsicElements, 'a' | 'button'>;
+export type AllowedElements = HTMLButtonElement | HTMLAnchorElement;
+export type ButtonElementType<Tag extends AllowedTags> = Tag extends 'a'
+  ? HTMLAnchorElement
+  : HTMLButtonElement;
+
+export type ButtonProps<Tag extends AllowedTags> = BaseButtonProps &
+  HTMLAttributes<Tag> &
+  JSX.IntrinsicElements[Tag] & {
+    ref?: Ref<ButtonElementType<Tag>>;
+  };
+
+function ButtonComponent<TagName extends AllowedTags>(
+  {
+    kind = ButtonKind.Button,
+    variant,
+    size = ButtonSize.Medium,
+    color,
+    className,
+    icon,
+    iconPosition = ButtonIconPosition.Left,
+    loading,
+    pressed,
+    children,
+    onClick,
+    ...props
+  }: ButtonProps<TagName>,
+  ref?: Ref<ButtonElementType<TagName>>,
+  // TODO: @milan - this is a hack, but only way how I could get TS to be type happy for now
+  Tag = kind === ButtonKind.Link ? 'a' : 'button',
+): ReactElement {
+  const iconOnly = icon && !children;
+  const getIconWithSize = useGetIconWithSize(size, iconOnly);
+  const isAnchor = kind === ButtonKind.Link;
+
+  return (
+    <Tag
+      {...props}
+      {...(isAnchor ? combinedClicks(onClick) : { onClick })}
+      aria-busy={loading}
+      aria-pressed={pressed}
+      ref={ref}
+      className={classNames(
+        { iconOnly },
+        size,
+        'btn flex flex-row items-center justify-center border typo-callout font-bold no-underline shadow-none cursor-pointer select-none focus-outline relative',
+        {
+          [`btn-${variant}`]: variant,
+          [`btn-${variant}-${color}`]: color && variant,
+        },
```

Comment:
Be aware this won't work.
[Tailwind](https://stackoverflow.com/a/76193787/1719919) doesn't allow mixed variables, you should always use fully qualified names.

Thus I think we need some mapping function that based on variant-color can return the final `btn-` class.

Reply:
I refactored it anyway, so that the code uses similar conventions as for button sizes. Please take a look again 👍